### PR TITLE
Switch to Flat encoding for [U]PLC terms.

### PR DESCRIPTION
**Pull request by raduom**
Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [x] `$(nix-build default.nix -A pkgsLocal.updateMaterialized)` to update the materialized Nix files
   - [x] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
   - [x] `$(nix-shell shell.nix --run fix-stylish-haskell)` to fix any formatting issues
- If you changed any Purescript files:
   - [ ] `$(nix-shell shell.nix --run fix-purty)` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-ledger/src/Ledger/Scripts.hs
Comment (raduom) [inline plutus-ledger/src/Ledger/Scripts.hs:97]: This instance will use the `Flat` encoding to generate a `ByteString` which then is encoded by `CBOR`. This is where we switch from PLC Flat to PAB CBOR.

**Comment by raduom** [comment]
I wanted some quick feedback on this. At the moment there are still some tests failing, so there is a bit more work that needs to be done, but everything compiles and most tests pass.

**Comment by michaelpj** [comment]
Switch to flat, surely? :upside_down_face:

**Comment by michaelpj** [comment]
(If it's not done I recommend making it a draft PR)

File: plutus-ledger/src/Ledger/Scripts.hs
Comment (michaelpj) [inline plutus-ledger/src/Ledger/Scripts.hs:89]: I think we probably *don't* want this: you shouldn't be using `Flat` to work with `Script`, you should be using `CBOR`.

File: plutus-ledger/src/Ledger/Scripts.hs
Comment (michaelpj) [inline plutus-ledger/src/Ledger/Scripts.hs:93]: Specifically: the ledger uses CBOR for transactions, so at some point it has to be CBOR.

Also, this means that the overall CBOR won't be totally "self-describing": it'll just say "here's a bytestring, no idea what's in there" unless you know it's a flat-encoded script. But that's fine.

TBH this is a bit tricky so probably worth a proper Note.

File: plutus-ledger/src/Ledger/Scripts.hs
Comment (michaelpj) [inline plutus-ledger/src/Ledger/Scripts.hs:129]: No need to change this, right? especially if `Script` keeps just its `Serialise` instance.

**Comment by raduom** [comment]
We need to decide if we will keep the CBOR instance. Based on this decision, here is how you should interpret this change.
1) If we remove CBOR, then we remove all comments.
2) If we keep CBOR, I will remove all commented code and make sure that everything has both CBOR and Flat instances.

File: plutus-ledger/src/Ledger/Scripts.hs
Comment (raduom) [reply plutus-ledger/src/Ledger/Scripts.hs:129]: Does this answer your question?
https://github.com/input-output-hk/plutus/pull/2482#issuecomment-731086632

Comment (michaelpj) [reply plutus-ledger/src/Ledger/Scripts.hs:129]: We should decide if we want to keep the CBOR instance for *Program*. We should *definitely* keep the CBOR instance for `Script`.

File: plutus-ledger/src/Ledger/Scripts.hs
Comment (raduom) [reply plutus-ledger/src/Ledger/Scripts.hs]: And it's gone.

**Comment by michaelpj** [comment]
I say keep the CBOR for now.

File: plutus-ledger/src/Ledger/Scripts.hs
Comment (raduom) [reply plutus-ledger/src/Ledger/Scripts.hs]: I have added a note.

File: plutus-benchmark/flat/Codec.hs
Comment (effectfully) [inline plutus-benchmark/flat/Codec.hs:8]: Commented out code.

File: plutus-core/plutus-ir/Language/PlutusIR.hs
Comment (effectfully) [inline plutus-core/plutus-ir/Language/PlutusIR.hs:73]: How come we now need `Flat fun`? `Datatype` is two `TyVarDecl`s (each doesn't reference `fun`) and a `VarDecl` (doesn't use the `fun` that it's parameterized by).

We should probably drop `fun` from `VarDecl`, I introduced it by mistake. I'll do that once this PR is merged. Pinging @michaelpj

**Review by effectfully — APPROVED** [review]
LGTM

File: plutus-benchmark/flat/Codec.hs
Comment (raduom) [reply plutus-benchmark/flat/Codec.hs]: Fixed.

File: plutus-core/plutus-ir/Language/PlutusIR.hs
Comment (michaelpj) [reply plutus-core/plutus-ir/Language/PlutusIR.hs:73]: Removing it from `VarDecl` seems fine. I vaguely remember we included it in some places we didn't need it because it helped type inference? Or am I makin that up?

**Review by michaelpj — APPROVED** [review]

File: plutus-core/plutus-ir/Language/PlutusIR.hs
Comment (effectfully) [reply plutus-core/plutus-ir/Language/PlutusIR.hs:73]: I'd included it in one place where it wasn't needed, but then we talked and I removed it. Not sure about `VarDecl`, this looks like an accidental change (I did `s/name name uni ann/name name uni fun ann`, i.e. whereever there's term-level names and universes, there's most likely built-in functions).

**Comment by michaelpj** [comment]
Okay, let's do this now, we can worry about the `fun` later.

**Pull request merged**
